### PR TITLE
Added JsonContractResolver to validate value type properties

### PR DESCRIPTION
**Pull request by kichalla**
@rynowak @harshgMSFT @yishaigalatzer

### CODE REVIEW

File: src/Microsoft.AspNet.Mvc.Core/Formatters/JsonOutputFormatter.cs [hunk 1]
```diff
@@ -22,6 +22,8 @@ public JsonOutputFormatter()
             SupportedMediaTypes.Add(MediaTypeHeaderValue.Parse("text/json"));
 
             _serializerSettings = new JsonSerializerSettings();
+
+            _serializerSettings.ContractResolver = new JsonContractResolver();
```
Comment (kichalla) [inline src/Microsoft.AspNet.Mvc.Core/Formatters/JsonOutputFormatter.cs]: FYI: With the current state of the contract resolver, it ideally applies only to the input side...but I applied it to output too as we plan to support camel case etc. in future in which case we need this on both input & output...

File: test/Microsoft.AspNet.Mvc.FunctionalTests/InputObjectValidationTests.cs [hunk 2]
```diff
@@ -175,8 +175,10 @@ var shouldNotContainMessage
             // Arrange
             var server = TestServer.Create(_services, _app);
             var client = server.CreateClient();
-            var content = new StringContent("{\"Alias\":\"xyz\"}", Encoding.UTF8, "application/json");
```
Comment (kichalla) [inline test/Microsoft.AspNet.Mvc.FunctionalTests/InputObjectValidationTests.cs]: Unrelated to this PR (more of a cleanup as this is clearly incorrect)

File: src/Microsoft.AspNet.Mvc.Core/Formatters/JsonContractResolver.cs [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+using System.ComponentModel.DataAnnotations;
+using System.Reflection;
+using Newtonsoft.Json;
+using Newtonsoft.Json.Serialization;
+
+namespace Microsoft.AspNet.Mvc
+{
+    /// <summary>
+    /// The default <see cref="IContractResolver"/> for <see cref="JsonInputFormatter"/> and 
+    /// <see cref="JsonOutputFormatter"/>.
```
Comment (yishaigalatzer) [inline src/Microsoft.AspNet.Mvc.Core/Formatters/JsonContractResolver.cs]: why is it  needed on a `JsonOutputFormatter` we are not validating on the way out?

File: src/Microsoft.AspNet.Mvc.Core/Formatters/JsonContractResolver.cs [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+    /// It determines if a value type member has <see cref="RequiredAttribute"/> and sets the appropriate 
+    /// JsonProperty settings.
+    /// </summary>
+    public class JsonContractResolver : DefaultContractResolver
+    {
+        protected override JsonProperty CreateProperty(MemberInfo member, MemberSerialization memberSerialization)
+        {
+            var property =  base.CreateProperty(member, memberSerialization);
+
+            var required = member.GetCustomAttribute(typeof(RequiredAttribute), inherit: true);
```
Comment (yishaigalatzer) [inline src/Microsoft.AspNet.Mvc.Core/Formatters/JsonContractResolver.cs]: is json.NET caching the results of this call?

File: src/Microsoft.AspNet.Mvc.Core/Formatters/JsonContractResolver.cs [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+            if (required != null)
+            {
+                var propertyType = ((PropertyInfo)member).PropertyType;
+
+                if (propertyType.IsValueType() && !propertyType.IsNullableValueType())
```
Comment (yishaigalatzer) [inline src/Microsoft.AspNet.Mvc.Core/Formatters/JsonContractResolver.cs]: add a comment explaining this check. How does this handle structs?

File: src/Microsoft.AspNet.Mvc.Core/Formatters/JsonOutputFormatter.cs (hunk 1 above)
Comment (yishaigalatzer) [inline src/Microsoft.AspNet.Mvc.Core/Formatters/JsonOutputFormatter.cs]: this is answering my comment above. But still don't apply to both sides. When we do camelcasing support we can reconsider, for now its not even on the list for the 6.0 milestone

File: src/Microsoft.AspNet.Mvc.Core/Formatters/JsonContractResolver.cs (hunk 5 above)
Comment (yishaigalatzer) [inline src/Microsoft.AspNet.Mvc.Core/Formatters/JsonContractResolver.cs]: I see the struct handling test below, it makes sense. I'd like to see a functional test for a type mixing struct with POCO and verify that the right modelstate errors are applied

**Comment by yishaigalatzer** [comment]
:watch:

File: src/Microsoft.AspNet.Mvc.Core/Formatters/JsonOutputFormatter.cs (hunk 1 above)
Comment (kichalla) [inline src/Microsoft.AspNet.Mvc.Core/Formatters/JsonOutputFormatter.cs]: sure...makes sense..will remove it from output formatter..

File: src/Microsoft.AspNet.Mvc.Core/Formatters/JsonContractResolver.cs (hunk 4 above)
Comment (kichalla) [inline src/Microsoft.AspNet.Mvc.Core/Formatters/JsonContractResolver.cs]: yes, right...for a property on a type, CreateProperty is called only once for the lifetime of the application...

File: src/Microsoft.AspNet.Mvc.Core/Formatters/JsonContractResolver.cs (hunk 5 above)
Comment (kichalla) [inline src/Microsoft.AspNet.Mvc.Core/Formatters/JsonContractResolver.cs]: sure...will do...

**Comment by kichalla** [comment]
Updated. @yishaigalatzer

**Comment by kichalla** [comment]
Ping

File: test/Microsoft.AspNet.Mvc.FunctionalTests/ModelBindingTest.cs [hunk 6]
```diff
@@ -1731,5 +1731,75 @@ public class ModelBindingTest
...
+            var data = await response.Content.ReadAsStringAsync();
+            var actualModelStateErrorMessages = JsonConvert.DeserializeObject<IEnumerable<string>>(data);
+            Assert.NotNull(actualModelStateErrorMessages);
+            Assert.Equal(expectedModelStateErrorMessages.Count(), actualModelStateErrorMessages.Count());
+            foreach (var expectedErrorMessage in expectedModelStateErrorMessages)
+            {
+                Assert.Contains(
+                    actualModelStateErrorMessages,
+                    (actualErrorMessage) => actualErrorMessage.StartsWith(expectedErrorMessage));
+
```
Comment (harshgMSFT) [inline test/Microsoft.AspNet.Mvc.FunctionalTests/ModelBindingTest.cs]: nit extra line

File: test/WebSites/ModelBindingWebSite/Controllers/ValidationController.cs [hunk 7]
```diff
@@ -21,6 +22,51 @@ public bool AvoidRecursive(SelfishPerson selfishPerson)
         {
             return ModelState.IsValid;
         }
+
+        public IActionResult RequiredAttributeOnValueAndReferenceTypes([FromBody] Rectangle drawing)
```
Comment (harshgMSFT) [inline test/WebSites/ModelBindingWebSite/Controllers/ValidationController.cs]: lets keep the action names more realistic(at least easy to read), the test names can depict what is being tested... I really like the models you selected and the way this is coded up, we could add comments if you want to explicitly clarify what is being tested here.

**Comment by harshgMSFT** [comment]
:shipit: once @yishaigalatzer is happy

**Comment by kichalla** [comment]
@yishaigalatzer : I updated the PR with a new test and comment. Please take a look.

File: test/Microsoft.AspNet.Mvc.FunctionalTests/InputObjectValidationTests.cs (hunk 2 above)
Comment (rynowak) [inline test/Microsoft.AspNet.Mvc.FunctionalTests/InputObjectValidationTests.cs]: :+1:

**Pull request merged**